homeassistant/components/opower/coordinator.py

```diff
@@ -86,6 +87,35 @@ async def _async_update_data(
         # Because Opower provides historical usage/cost with a delay of a couple of days
         # we need to insert data into statistics.
         await self._insert_statistics()
+        # Estimate forecast values with the last known rate if they are empty
+        for _, forecast in enumerate(forecasts):
+            if (
+                forecast.usage_to_date > 0
+                and forecast.cost_to_date == 0
+                and forecast.account.utility_account_id in self.last_known_rates
+            ):
+                forecast.cost_to_date = (
+                    self.last_known_rates[forecast.account.utility_account_id]
```

Comment:
I could take out the forecast entity support, or make it configurable or something, but I'm not sure what would be done about the recent energy panel data. If the estimated rate was exposed somewhere, it would be trivial to calculate with an automation, but I'm not sure how you would go about calculating that rate in the first place. I wasn't able to find anything about accessing historical energy data in HA.

There's the cumulative entities, but the rate would decrease over the course of the month, and spike up again at the next billing period when the old costs are updated. It wouldn't be significant, but also the rate would be even less accurate, as it could have changed significantly over the years.

The issue with PSE is there's no way to programmatically obtain their rates. I'm not sure about PG&E, but PSE only publishes PDFs once a month, which would be a huge pain to process. The easiest, good enough solution right now is just to make a best guess.

Reply:
In the past I received this feedback from a home assistant core member "We normally don't allow device or service integrating integrations to calculate state. They should represent the integration api data as is in accordance with our entity design."

Here you violate this.

Regardless, this is an approximation since it can't take into account basic charges, tiered charges, change of season charges, time of use, distribution charges etc. that vary, so it will only confuse users who will complain the costs don't match to what they see in their bill.

Reply:
That makes sense, but is there another way to present the billing cost data for more recent periods that get overwritten by the daily/hourly views?

Just to be clear, unless I misunderstood what you meant, are you talking about the forecast estimates that the utility does not provide, or are you talking about the past data that was spread out? For the past data, I would argue it's more annoying and inconsistent that I can see the billing costs 3 years back, but I can't see it the moment it switches to daily data. To me at least, that behavior is far more confusing than the billing costs being a few cents to maybe a dollar off from the cumulative errors. And with that, the usage data is already incorrect on the website/HA anyways because of small rounding errors, so I don't think the cost issue is significant there. That's why I figured simply spreading the charge across the window is good enough for utilities that don't have such fine grained data to begin with. My goal was that it shouldn't interfere with any utility that properly provides finer cost breakdowns too.

If you were speaking about the estimated forecasts though, I understand how that can go against the design principles. If there was some way to calculate this in an automation, or to expose some extra data, perhaps as a default disabled entity, that could be used to calculate it, that'd probably be the best solution for me, but if there's no good way that's understandable.

Reply:
I was speaking about both the forecast sensors and the statistics.

> I would argue it's more annoying and inconsistent that I can see the billing costs 3 years back, but I can't see it the moment it switches to daily data.

Yes I'm aware of this issue for utilities like yours. I'm OK approving a PR that fixes that without estimating costs as long as it's not too complex. There are 2 separate statistics so for utilities like yours we could store the cost data monthly while the usage hourly.

> If there was some way to calculate this in an automation

At least for the forecast sensors you should be able to use the sql integration to get the previous month rate in a sensor and multiply it with the usage sensor in a template sensor.

Reply:
Thanks, I'll take a look at that. I didn't know HA is able to deal with the cost and usage being completely separate.

Reply:
I'm not sure this is a very good solution. If the cost and consumption data windows don't match up, HA simply dumps the full cost into the first matching window. Scrolling through the data, everything is $0 until the day the billing cycle started, where the full cycle cost shows up. Perhaps it would be better if the window is moved so that the cost shows up at the end, when you're actually charged? But either way, it's still just confusing and hard to use. It's still useful when looking at big time windows across multiple months/years though.

Anything is definitely better than before, but I'm still of the opinion that spreading the cost across the billing period makes more sense, even though the hourly costs might not be perfect if you have variable pricing.

I put it at [opower_split_cost_consumption](https://github.com/home-assistant/core/compare/dev...PseudoResonance:homeassistant-core:opower_split_cost_consumption) if you want to take a look.